components/org.wso2.carbon.identity.recovery/src/main/java/org/wso2/carbon/identity/recovery/handler/UserSelfSignUpHandler.java

```diff
@@ -0,0 +1,232 @@
+                    }
+
+                    UserRecoveryDataStore userRecoveryDataStore = JDBCRecoveryDataStore.getInstance();
+                    userRecoveryDataStore.invalidateByUserUniqueId(user.getUniqueUserId());
+
+                    String secretKey = Utils.generateUUID();
+                    UserRecoveryData recoveryDataDO = new UserRecoveryData(user.getUniqueUserId(), secretKey,
+                                                                           RecoveryScenarios.SELF_SIGN_UP,
+                                                                           RecoverySteps.CONFIRM_SIGN_UP);
+                    userRecoveryDataStore.store(recoveryDataDO);
+
+                    if (config.isNotificationInternallyManaged()) {
+
+                        Property[] arbitraryProperties = (Property[]) eventProperties.get(SELF_SIGN_UP_PROPERTIES);
+                        Utils.triggerNotification(eventService, user.getUniqueUserId(),
+                                                  NOTIFICATION_TYPE_ACCOUNT_CONFIRM,
+                                                  secretKey,
+                                                arbitraryProperties);
+                    } else {
+                        eventContext.addParameter(IdentityStoreConstants.UNIQUE_USED_ID, user.getUniqueUserId());
+                        eventContext.addParameter(CONFIRMATION_CODE, secretKey);
+                    }
+                } else {
+
+                    if (log.isDebugEnabled()) {
+                        log.debug("Changing user lifecycle state to {} for user: {}", UserState
+                                .UNLOCKED__UNVERIFIED.toString(), user.getUniqueUserId());
+                    }
+                    LifecycleOperationManager.executeLifecycleEvent(UserState.UNLOCKED__UNVERIFIED.toString(),
+                                                                    domainUID, domainUID, null);
+                    identityStore.setUserState(user.getUniqueUserId(), UserState.UNLOCKED__UNVERIFIED.toString());
+                }
+            } catch (LifecycleException e) {
+                throw Utils.handleServerException(ErrorCodes.FAILED_LIFECYCLE_EVENT, null, e);
+            } catch (UserNotFoundException e) {
+                throw Utils.handleServerException(ErrorCodes.USER_NOT_FOUND, null, e);
+            } catch (IdentityStoreException e) {
+                throw Utils.handleServerException(ErrorCodes.FAILED_USER_STATE_UPDATE, null, e);
+            }
+        }
+
+        if (IdentityStoreInterceptorConstants.POST_AUTHENTICATE.equals(event.getEventName())) {
+
+            Map<String, Object> eventProperties = event.getEventProperties();
+
+            AuthenticationContext authenticationContext = (AuthenticationContext) eventProperties.get
+                    (IdentityStoreConstants.AUTHENTICATION_CONTEXT);
+
+            if (authenticationContext.isAuthenticated()) {
+
+                User user = authenticationContext.getUser();
```

Comment:
Why we check isAuthenticated() to get User, and throw exception when LOCKED_SELF_SIGN_UP ? . Yes, it is true that authenticationContext.getUser(); is not null when authenticationContext.isAuthenticated() == true. However in coding it is not related. the better way would be, irrespective of authenticated or not, since this is post auth handler. 
User user = authenticationContext.getUser()
if(user != null && UserState.LOCKED_SELF_SIGN_UP.toString().equals(user.getState())) { .. do stuff}